File: 2021/10/17/unique_emails.py
"""
https://leetcode.com/problems/unique-email-addresses/
Author: Andrew Jarombek
Date: 10/17/2021
"""

from typing import List


def num_unique_emails(emails: List[str]) -> int:
    addresses = set()

    for email in emails:
        index = 0

        address = ''
        plus_section = False

        while index < len(email):
            if email[index] == '@':
                address += email[index:]
                index = len(email)
            elif email[index] == '+':
                plus_section = True
                index += 1
            elif not plus_section and email[index] != '.':
                address += email[index]
                index += 1
            else:
                index += 1

        addresses.add(address)

    return len(addresses)

File: 2023/05/18/uniqueEmails.js
/**
 * https://leetcode.com/problems/unique-email-addresses/
 * @author Andrew Jarombek
 * @since 5/18/2023
 */

const assert = require('assert');

// replaceAll() not available in older JS versions
String.prototype.replaceAll = function (search, replacement) {
    return this.split(search).join(replacement);
};

const numUniqueEmails = function(emails) {
    const set = new Set();

    for (let email of emails) {
        const [fullLocalName, domainName] = email.split('@');
        let localName = fullLocalName.split('+')[0];
        localName = localName.replaceAll('.', '');

        set.add(`${localName}@${domainName}`);
    }

    return set.size;
};

// Test case 1: Unique emails
let emails1 = [
    "test.email@gmail.com",
    "test.e.mail+john.smith@gmail.com",
    "testemail@gmail.com"
];
assert.strictEqual(numUniqueEmails(emails1), 1, 'Test case 1 failed');

// Test case 2: Duplicate emails
let emails2 = [
    "test.email@gmail.com",
    "test.email@yahoo.com",
    "test.email@gmail.com",
    "test.e.mail+john.smith@gmail.com"
];
assert.strictEqual(numUniqueEmails(emails2), 2, 'Test case 2 failed');
